frontend/src/components/resourceMap/GraphView.tsx

```diff
@@ -0,0 +1,390 @@
+        return;
+      }
+      _setSelectedNodeId(id);
+    },
+    [_setSelectedNodeId]
+  );
+
+  // Expand all groups state
+  const [expandAll, setExpandAll] = useState(false);
+
+  // Load source data
+  const { nodes, edges, selectedSources, sourceData, toggleSelection } = useSources();
+
+  // Graph with applied layout, has sizes and positions for all elements
+  const [layoutedGraph, setLayoutedGraph] = useState<{ nodes: Node[]; edges: Edge[] }>({
+    nodes: [],
+    edges: [],
+  });
+
+  const flow = useReactFlow();
+
+  // Apply filters
+  const filteredGraph = useMemo(() => {
+    const filters = [...defaultFilters];
+    if (hasErrorsFilter) {
+      filters.push({ type: 'hasErrors' });
+    }
+    if (namespaces) {
+      filters.push({ type: 'namespace', namespaces });
+    }
+    return filterGraph(nodes, edges, filters);
+  }, [nodes, edges, hasErrorsFilter, namespaces, defaultFilters]);
+
+  // Group the graph
+  const { visibleGraph, fullGraph } = useMemo(() => {
+    const graph = groupGraph(filteredGraph.nodes as KubeObjectNode[], filteredGraph.edges, {
+      groupBy,
+    });
+
+    const visibleGraph = collapseGraph(graph, { selectedNodeId, expandAll }) as GroupNode;
+
+    return { visibleGraph, fullGraph: graph };
+  }, [filteredGraph, groupBy, selectedNodeId, expandAll]);
+
+  // Apply layout to visible graph
+  const aspectRatio = useStore(it => it.width / it.height);
+  useEffect(() => {
+    applyGraphLayout(visibleGraph, aspectRatio).then(layout => {
+      setLayoutedGraph(layout);
+      setTimeout(() => {
+        const bounds = flow.getNodesBounds(layout.nodes);
+
+        // Only fit bounds when user hasn't moved viewport manually
+        if (!viewportMovedRef.current) {
+          flow.fitBounds(bounds, { padding: 0.2 });
+        }
+      });
+    });
+  }, [visibleGraph, aspectRatio, flow]);
+
+  // Reset after view change
+  useLayoutEffect(() => {
+    viewportMovedRef.current = false;
+  }, [selectedNodeId, groupBy, expandAll]);
+
+  const selectedNode = nodes.find((it: GraphNode) => it.id === selectedNodeId);
+
+  let selectedGroup: KubeGroupNode | undefined;
+  if (selectedNodeId) {
+    selectedGroup = findGroupContaining(visibleGraph, selectedNodeId) as any;
+  }
+  const highlights = useNodeHighlight(selectedNodeId);
+
+  const graphSize = getGraphSize(visibleGraph);
+  useEffect(() => {
+    if (expandAll && graphSize > 50) {
+      setExpandAll(false);
+    }
+  }, [graphSize]);
+
+  return (
+    <GraphViewContext.Provider
+      value={useMemo(
+        () => ({ nodeSelection: selectedNodeId, highlights, setNodeSelection: setSelectedNodeId }),
+        [selectedNodeId, setSelectedNodeId, highlights]
+      )}
+    >
+      <Box
+        sx={{
+          position: 'relative',
+          height: height ?? '800px',
+          display: 'flex',
+          flexDirection: 'row',
+          flex: 1,
+        }}
+      >
+        <CustomThemeProvider>
+          <Box
```

Comment:
This whole header maybe should be abstracted, if you don't have to pass over a 100 props.

Reply:
there's a lot of props that'd need to be passed. I agree that this part would make sense as a separate component but I think for now this is okay

Reply:
Let's skip this. J is away today, so I'll resolve.